src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -1011,6 +1062,271 @@ trait Implicits {
+      i.tpe match {
+        case PolyType(args, mt) =>
+          val tp = mt.finalResultType
+          i.tpe.paramLists match {
+            case List(params) if params.head.isImplicit => !params.exists(param => recursive(param.info, tp))
+            case Nil => true
+            case _ => false
+          }
+
+        case NullaryMethodType(_) => true
+        case other => false
+      }
+
+    def isInduction(i: ImplicitInfo): Boolean =
+      i.tpe match {
+        case PolyType(args, mt) =>
+          val tp = mt.finalResultType
+          i.tpe.paramLists match {
+            case List(params) if params.head.isImplicit =>
+              params.filter(param => recursive(param.info, tp)) match {
+                case List(rec) => convergent(rec.info, tp)
+                case _ => false
+              }
+            case _ => false
+          }
+        case _ => false
+      }
+
+    def inductive(local: List[ImplicitInfo], scope: List[ImplicitInfo]): Boolean = {
+      val exhaustive =
+        (local.exists(isBase) || scope.exists(isBase)) &&
+        (local.exists(isInduction) || scope.exists(isInduction)) &&
+        local.forall(i => !i.sym.isMacro && (isBase(i) || isInduction(i))) &&
+        scope.forall(i => !i.sym.isMacro && (isBase(i) || isInduction(i)))
+
+      if(exhaustive) {
+        val inductions = local.filter(isInduction) ++ scope.filter(isInduction)
+        inductions.map(divergenceMap).transpose.forall { convergences =>
+          !(convergences.exists(_ == Convergent) && convergences.exists(_ == Divergent))
+        }
+      } else false
+    }
+
+    def solve(pt: Type, tp: Type): SearchResult = {
+      val wildPt = deriveTypeWithWildcards(context.outer.undetparams ++ context.undetparams)(pt)
+      val (tp0, tparams) = tp match {
+        case PolyType(tparams, restpe) => (ApproximateDependentMap(restpe).finalResultType, tparams)
+        case _ => (ApproximateDependentMap(tp).finalResultType, Nil)
+      }
+
+      val allUndetparams = (context.outer.undetparams ++ context.undetparams ++ tparams).distinct
+      val tvars = allUndetparams map freshVar
+      val tpInstantiated = tp0.instantiateTypeParams(allUndetparams, tvars)
+
+      if(!matchesPt(tpInstantiated, wildPt, allUndetparams)) SearchFailure
+      else {
+        val targs = solvedTypes(tvars, allUndetparams, allUndetparams map varianceInType(pt), upper = false, lubDepth(wildPt :: tp :: Nil))
+        if(!checkBounds(EmptyTree, NoPrefix, NoSymbol, allUndetparams, targs, "inferred ")) SearchFailure
+        else {
+          val AdjustedTypeArgs(okParams, okArgs) = adjustTypeArgs(allUndetparams, tvars, targs)
+          val undetparams = allUndetparams diff okParams
+          val tpSubst = tp0.instantiateTypeParams(okParams, okArgs)
+          if(!matchesPt(tpSubst, wildPt, undetparams)) SearchFailure
+          else {
+            val subst: TreeTypeSubstituter =
+              if (okParams.isEmpty) EmptyTreeTypeSubstituter
```

Comment:
Agreed. My normal style is no space, the scalac style is a space. I try to conform to local customs but sometime slip.